src/download_target.cpp

```diff
@@ -34,11 +35,55 @@ namespace powerloader
         if (m_is_zchunk)
         {
             m_p_zck = std::make_unique<zck_target>();
-            m_p_zck->zck_cache_file = filename;
+            m_p_zck->zck_cache_file = m_destination_path;
         }
 #endif
     }
 
+    std::shared_ptr<DownloadTarget> DownloadTarget::from_url(Context& ctx,
+                                                             const std::string& target_url,
+                                                             const fs::path& destination_path,
+                                                             const fs::path& destination_dir)
+    {
+        if (contains(target_url, "://"))
+        {
+            // even when we get a regular URL like `http://test.com/download.tar.gz`
+            // we want to create a "mirror" for `http://test.com` to make sure we correctly
+            // retry and wait on mirror failures
+            URLHandler uh{ target_url };
+            const std::string url = uh.url();
+            const std::string host = uh.host();
+            const std::string path = uh.path();
+            const std::string mirror_url = url.substr(0, url.size() - path.size());
+            const fs::path dst = destination_path.empty() ? fs::path{ rsplit(path, "/", 1).back() }
+                                                          : destination_path;
+
+            // Note that `ctx.mirror_map[host]` will create the vector by default at key `host` if
+            // it doesnt already exists.
+            ctx.mirror_map[host].push_back(std::make_shared<Mirror>(ctx, mirror_url));
```

Comment:
shouldn't we check if the mirror already exists?

Reply:
OK, got it, sorry!

Reply:
Yeah this code already makes sure there is a vector created.

However if you meant check that in the vector there isnt already a mirror witht the same `mirror_url`, I agree! I was thinking about this while writing this. I'm considering replacing the vector by a `flat_set` (or equivalent) (basically a  vector disguised as a set, only accepts one instance of each value, interface of a set, but still contiguous (and ordered))

Reply:
unresolved to remind me to do this